cpp/src/bitmask/bit_mask.cuh

```diff
@@ -14,143 +14,224 @@
+    RMM_TRY(RMM_FREE(valid, 0));
+    return GDF_SUCCESS;
   }
 
   /**
-   * @brief Set a bit in a thread-safe manner
+   *  @brief Get a single element of bits from the device
+   *
+   *  @param[out]  element - address on host where the bits will be stored
+   *  @param[out]  device_element - address on the device containing the bits to fetch
    *
-   * @param[in] record_idx    the record index
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
    */
-  template <typename T>
-  __device__ void set_bit(T record_idx) {
-    bit_mask::set_bit_safe(valid_, record_idx);
+  inline gdf_error get_element(bit_mask_t *element, const bit_mask_t *device_element) {
+    CUDA_TRY(cudaMemcpy(element, device_element, sizeof(bit_mask_t), cudaMemcpyDeviceToHost));
+    return GDF_SUCCESS;
   }
 
-
   /**
-   * @brief Clear a bit in a thread-safe manner
+   *  @brief Put a single element of bits to the device
    *
-   * @param[in] record_idx    the record index
+   *  @param[out]  element - address on host containing bits to store
+   *  @param[out]  device_element - address on the device where the bits will be stored
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
    */
-  template <typename T>
-  __device__ void clear_bit(T record_idx) {
-    bit_mask::clear_bit_safe(valid_, record_idx);
+  inline gdf_error put_element(bit_mask_t element, bit_mask_t *device_element) {
+    CUDA_TRY(cudaMemcpy(device_element, &element, sizeof(bit_mask_t), cudaMemcpyHostToDevice));
+    return GDF_SUCCESS;
   }
 
   /**
-   * @brief Get the number of elements in this bit container
+   *  @brief Allocate device space for the valid bit mask.
    *
-   * @return the number of elements
+   *  @param[out] mask                  address of the bit mask pointer
+   *  @param[in]  number_of_records     number of records
+   *  @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other
+   *                                    values indicate un-initialized.  Default is uninitialized
+   *  @param[in]  padding_bytes         optional, specifies byte boundary the data should be padded to.
+   *                                    Defaults to 64 bytes, meaning the space allocated will be rounded
+   *                                    up to the next multiple of 64 bytes.
+   *
+   *  @return GDF_SUCCESS on success, the RMM or CUDA error on error
    */
-  __device__ gdf_size_type num_elements() const {
-    return bit_mask::num_elements(bitlength_);
+  gdf_error create_bit_mask(bit_mask_t **mask, gdf_size_type number_of_records, int fill_value = -1, gdf_size_type padding_bytes = 64) {
+    //
+    //  To handle padding, we will round the number_of_records up to the next padding boundary, then identify how many element
+    //  that equates to.  Then we can allocate the appropriate amount of storage.
+    //
+    gdf_size_type num_bytes = (number_of_records + 7) / 8;
```

Comment:
Maybe
Hmmm, maybe the following?  Although the original is much more concise...

```suggestion
    gdf_size_type num_bytes = cudf::util::div_rounding_up_unsafe<gdf_size_type>(number_of_records, CHAR_BIT);
```